auth/src/source.rs

```diff
@@ -0,0 +1,474 @@
+            serde_json::from_slice(contents)?
+        } else if let Some(file_path) = self.file_path {
+            serde_json::from_slice(&std::fs::read(file_path)?)?
+        } else {
+            return Err(Error::Other("".into()));
+        };
+        if self.scopes.is_empty() {
+            return Err(Error::Other("".into()));
+        }
+        Ok(ServiceAccountKeySource {
+            file: sa,
+            scopes: self.scopes,
+        })
+    }
+}
+
+/// A [Source] derived from a Service Account Key.
+pub struct ServiceAccountKeySource {
+    file: ServiceAccountKeyFile,
+    scopes: Vec<String>,
+}
+
+/// A representation of a Service Account File.
+#[derive(Deserialize)]
+struct ServiceAccountKeyFile {
+    #[serde(rename = "type")]
+    cred_type: String,
+    client_email: String,
+    private_key_id: String,
+    private_key: String,
+    auth_uri: String,
+    token_uri: String,
+    project_id: String,
+}
+
+impl ServiceAccountKeySource {
+    pub fn from_file(path: impl AsRef<Path>) -> ServiceAccountKeySourceBuilder<'static> {
+        ServiceAccountKeySourceBuilder {
+            file_path: Some(path.as_ref().to_owned()),
+            contents: None,
+            scopes: Vec::new(),
+        }
+    }
+    pub fn from_file_contents(contents: &[u8]) -> ServiceAccountKeySourceBuilder {
+        ServiceAccountKeySourceBuilder {
+            file_path: None,
+            contents: Some(contents),
+            scopes: Vec::new(),
+        }
+    }
+
+    async fn _access_token(&self) -> Result<AccessToken> {
+        let signer = self.signer()?;
+        let payload = self.payload(signer)?;
+        let client = reqwest::Client::new();
+        let res = client
+            .post(self.file.token_uri.as_str())
+            .form(&ServiceAccountTokenRequest {
+                grant_type: DEFAULT_OAUTH_GRANT.into(),
+                assertion: payload,
+            })
+            .send()
+            .await
+            .map_err(|_| Error::Other("unable to make request to oauth endpoint".into()))?;
+        if !res.status().is_success() {
+            return Err(Error::Other(format!(
+                "bad request with status: {}",
+                res.status()
+            )));
+        }
+        let token_response: TokenResponse = res
+            .json()
+            .await
+            .map_err(|_| Error::Other("unable to decode response".into()))?;
+
+        Ok(AccessToken {
+            value: token_response.access_token,
+            expires: Some(Utc::now() + Duration::seconds(token_response.expires_in)),
+        })
+    }
+
+    fn signer(&self) -> Result<Box<dyn Signer>> {
+        let pk = rustls_pemfile::read_one(&mut self.file.private_key.as_bytes())?
+            .ok_or_else(|| Error::Other("".into()))?;
+        let pk = match pk {
+            Item::RSAKey(item) => item,
+            Item::PKCS8Key(item) => item,
+            _ => return Err(Error::Other("expected key in a different format".into())),
```

Comment:
In the error message, list the format provided and the ones that would have been accepted.

Reply:
done.